synapse/tools/migrate_010.py

```diff
@@ -0,0 +1,714 @@
+                start = oldv
+                tags[oldk[2:]] = (start, end)
+            elif oldk[0] == '<':
+                start, end = tags.get(oldk[2:], (None, None))
+                end = oldv
+                tags[oldk[2:]] = (start, end)
+            elif oldk == formname:
+                pk = self.convert_primary(oldprops)
+            elif oldk in self.subs:
+                # Skip if a sub to a comp or sepr that's another secondary property
+                continue
+            elif 'defval' in propmeta and oldv == propmeta['defval']:
+                # logger.debug('Skipping field %s with default value', oldk)
+                continue
+            elif oldk.startswith(formname + ':'):
+                if oldk[len(formname) + 1:] in ('seen:min', 'seen:max'):
+                    new_pname, newv = self.handle_seen(oldk, oldv, props)
+                elif formname in self.xrefs:
+                    # skip all non-seed secondary props in xrefs
+                    continue
+                else:
+                    new_pname, newv = self.convert_subprop(formname, oldk, oldv, oldprops)
+                props[new_pname] = newv
+            elif oldk == 'node:created':
+                props['.created'] = oldv
+            else:
+                # logger.debug('Skipping propname %s', oldk)
+                pass
+        if pk is None:
+            raise ConsistencyError('Missing pk', oldprops)
+
+        if formname == 'file:txtref':
+            props['type'] = 'text'
+        elif formname == 'file:imgof':
+            props['type'] = 'image'
+        elif formname == 'ps:image':
+            props['type'] = 'image'
+
+        if formname in self.form_renames:
+            formname = self.form_renames[formname]
+
+        assert '' not in props
+
+        retn = ((formname, pk), {'props': props})
+        if tags:
+            retn[1]['tags'] = tags
+
+        return retn
+
+    def convert_ps_image(self, formname, props):
+        '''
+        Transform ps:image into a file:ref
+        '''
+        person, file = self.convert_sepr(formname, formname, props[formname], props)
+        return (file, ('ps:person', person))
+
+    primary_prop_special = {
+        'inet:web:post': just_guid,
+        'it:dev:regval': just_guid,
+        'file:bytes': convert_file_bytes,
+        'inet:udp4': convert_inet_xxp_primary,
+        'inet:udp6': convert_inet_xxp_primary,
+        'inet:tcp4': convert_inet_xxp_primary,
+        'inet:tcp6': convert_inet_xxp_primary,
+        'ps:image': convert_ps_image,
+        'file:base': check_file_base_primary,
+    }
+
+    def handle_seen(self, propname, propval, newprops):
+        '''
+        :seen:min, :seen:max -> .seen = (min, max)
+        '''
+        seenmin, seenmax = newprops.get('.seen', (None, None))
+        if propname.endswith(':min'):
+            seenmin = propval
+        if propname.endswith(':max'):
+            seenmax = propval
+        if seenmax is None:
```

Comment:
This could break. We probably want to check `if seenmax is None and seenmin is not None`, otherwise you could end up doing `None + 1`

Reply:
The only way that is possible is if propval is None, which is impossible I think.

Either seenmin or seenmax are set. This sets the other.